src/ripple/app/consensus/RCLValidations.cpp

```diff
@@ -0,0 +1,254 @@
+        currentStale.reserve(512);
+        staleValidations_.swap(currentStale);
+
+        {
+            ScopedUnlockType sul(staleLock_);
+            {
+                auto db = app_.getLedgerDB().checkoutDb();
+
+                Serializer s(1024);
+                soci::transaction tr(*db);
+                for (auto const& rclValidation : currentStale)
+                {
+                    STValidation::pointer const& val = rclValidation.unwrap();
+                    val->add(s);
+
+                    auto const ledgerHash = to_string(val->getLedgerHash());
+
+                    boost::optional<std::uint64_t> ledgerSeq;
+                    *db << findSeq, soci::use(ledgerHash),
+                        soci::into(ledgerSeq);
+
+                    auto const initialSeq = ledgerSeq.value_or(
+                        app_.getLedgerMaster().getCurrentLedgerIndex());
+                    auto const nodePubKey = toBase58(
+                        TokenType::TOKEN_NODE_PUBLIC, val->getSignerPublic());
+                    auto const signTime =
+                        val->getSignTime().time_since_epoch().count();
+
+                    soci::blob rawData(*db);
+                    rawData.append(
+                        reinterpret_cast<const char*>(s.peekData().data()),
+                        s.peekData().size());
+                    assert(rawData.get_len() == s.peekData().size());
+
+                    *db << insVal, soci::use(initialSeq), soci::use(ledgerSeq),
+                        soci::use(ledgerHash), soci::use(nodePubKey),
+                        soci::use(signTime), soci::use(rawData);
+                }
+
+                tr.commit();
+            }
+        }
+    }
+
+    staleWriting_ = false;
+}
+
+bool
+handleNewValidation(Application& app,
+    STValidation::ref val,
+    std::string const& source)
+{
+    PublicKey const& signer = val->getSignerPublic();
+    uint256 const& hash = val->getLedgerHash();
+
+    // Ensure validation is marked as trusted if signer currently trusted
+    boost::optional<PublicKey> pubKey = app.validators().getTrustedKey(signer);
+    if (!val->isTrusted() && pubKey)
+        val->setTrusted();
+    RCLValidations& validations  = app.getValidations();
+
+    beast::Journal j = validations.journal();
+
+    if (!val->isTrusted())
+    {
+        JLOG(j.trace()) << "Node "
+                        << toBase58(TokenType::TOKEN_NODE_PUBLIC, signer)
+                        << " not in UNL st="
+                        << val->getSignTime().time_since_epoch().count()
+                        << ", hash=" << hash
+                        << ", shash=" << val->getSigningHash()
+                        << " src=" << source;
+    }
+
+    // If not currently trusted, see if signer is currently listed
+    if (!pubKey)
+        pubKey = app.validators().getListedKey(signer);
+
+    bool shouldRelay = false;
+
+    // only add trusted or listed
+    if (val->isTrusted() || pubKey)
```

Comment:
@wilsonianb This was my approach to eliminating the extra `isCurrent` call.  Can you spot check it?